src/device/device_ledger.cpp

```diff
@@ -1080,25 +1082,24 @@ namespace hw {
 
         //derivattion data
         memmove(derivation.data, &this->buffer_recv[0], 32);
-
-        #ifdef DEBUG_HWDEVICE
-        crypto::key_derivation derivation_clear  = hw::ledger::decrypt(derivation);
-        hw::ledger::check32("generate_key_derivation", "derivation", derivation_x.data, derivation_clear.data);
-        #endif
-
-        unlock_device();
-      }catch (...)  {
-        unlock_device();
-        throw;
+        r = true;
       }
-      return true;
+      #ifdef DEBUG_HWDEVICE
+      crypto::key_derivation derivation_clear ;
+        if ((this->mode == TRANSACTION_PARSE)  && has_view_key) {
+          derivation_clear  = derivation;
+        }else {
+          derivation_clear  = hw::ledger::decrypt(derivation);
+        }
+      hw::ledger::check32("generate_key_derivation", "derivation", derivation_x.data, derivation_clear.data);
+      #endif
+
+      return r;
     }
 
     bool device_ledger::derivation_to_scalar(const crypto::key_derivation &derivation, const size_t output_index, crypto::ec_scalar &res) {
```

Comment:
Doesn't this function also require the same branch logic with `has_view_key`?

Reply:
No (if I correctly analyse the code). Once the TX identified, the whole process is done throw device, even the re-computation of `derivation` encrypted from device

Reply:
OK, I'm fine if it's confirmed to work:)